### ISSUE TITLE
tz - with invalid (semivalid) format str, date::parse may fail to set istream in fail() condition and may return incorrect timepoints

### ISSUE THREAD

**Issue by acolomitchi** [HowardHinnant/date#79]
The documentation lead me to believe the `date::parse` would accept `strftime` format strings <b>**</b>, thus is stumbled over this bug trying to use `"%FT%T%Z"` as one.<br>My initial surprise was the returned value - doesn't contains what one expected.  After reading (<i>far down on the page</i>) the spec for `date::parse`, I learned the function actually uses `time_get`-style of format string; a check on the `time_get::get()` format string spec reveals there's no `%F` (shortcut for `%Y-%m-%d`) and then my surprise changed into  "Why no sign of failure and a bad result?"

Config: `g++ (Ubuntu 5.4.1-2ubuntu1~14.04) 5.4.1 20160904`

Code to repro:

<pre>
#include &lt;iostream>
#include &lt;locale>
#include "date/tz.h"

int main() {
  namespace d=date;
  std::istringstream input;
  std::string toParse, fmt;

  std::cout << std::endl << "date::parse with invalid format (failure expected)" << std::endl;
  toParse="2016-09-09T08:45:05Z+1000";
  fmt="%FT%T%Z";
  input.str(toParse);
  input.clear();
  d::sys_time<std::chrono::milliseconds> parsedDate;
  d::parse(input, fmt.c_str(), parsedDate);
  std::cout << "Using input: `" << toParse << "` with a format of `"
            << fmt << "` (failure expected)" << std::endl
  ;
  if(input.fail()) {
    std::cout << "Failed date::parse with input date:`"<< toParse <<"`"
        << "` using format `" <<  fmt << "`"
        << std::endl
    ;
  }
  else {
    std::cout << "Parse success; applying date::make_zoned gets `"
              <<  d::make_zoned("Australia/Melbourne", parsedDate)
              <<  "`" << std::endl;
  }
  // Bug already demonstrated by the code above
  // Below is an attempt to see if my expectations (of having istream.fail()-ing)
  // are unreasonable.
  // So...
  // how does time_get behaves?
  toParse="2016-09-09T08:45:05"; // doh. Of course...
  fmt="%FT%T";                   // ...  there's no timezone
  std::cout << "\n---\nAttempting the behaviour of time_get with invalid format `"
            <<  fmt << "` (expecting failure) " << std::endl
  ;
  input.str(toParse);
  input.clear();
  auto& tget=std::use_facet&lt;std::time_get&lt;char>>(input.getloc());
  struct tm tgetResult;
  std::ios_base::iostate err = std::ios_base::goodbit;
  tget.get(
      std::istreambuf_iterator<char>(input),  std::istreambuf_iterator<char>(),
      input,  err,  &tgetResult,
      fmt.c_str(),  fmt.c_str()+fmt.length()
  );
  input.setstate(err);
  if(input.fail()) {
    std::cout << "Failed time_get parsing of `" << toParse
              << "` using format `" << fmt << "`"
              << std::endl
    ;
  }
  else {
    char buff[256];
    strftime(buff, sizeof(buff), fmt.c_str(), &tgetResult);
    std::cout << "Parse success; applying strftime gets `"
              << buff << "`"
              << std::endl
    ;
  }
  return 0;
}
</pre>

-----
<b>**</b> May I kindly ask you to adjust the doc on [tz](https://howardhinnant.github.io/date/tz.html) so that any reference to `strftime` is replaced by `time_get / time_put`? The spec of `date::parse` is so buried down the page that anyone reading the doc is likely to believe the  `strftime` format string is applicable.

**Comment by acolomitchi** [comment]
The bug is located in the function of:
<pre>
template &lt;class CharT, class Traits, class Duration>
void
parse(std::basic_istream<CharT, Traits>& is,
      const std::basic_string<CharT, Traits>& format, sys_time<Duration>& tp,
      std::basic_string<CharT, Traits>& abbrev, std::chrono::minutes& offset)
</pre>
(currently at `date.h` line 4129)

All the calls into `f.get(...)` (with `decltype(f)==std::time_get<CharT>`) **do not check** if any error condition is signalled into the `err` variable once the call is finished. As such, the `parse` function remain oblivious to any error in the format string.

**Comment by acolomitchi** [comment]
PR #80 created containing the proposed fix.

Note: I'm not familiar with the testing procedures for this project, otherwise I'd create a regression test case for it. Info && / || references about testing will be welcomed

-

**Comment by HowardHinnant** [comment]
Thanks for the detailed bug report and fix.  I will review it in more detail later today.

Meanwhile I did not realize that `%F` was not part of the `get_time` spec.  libc++ accepts it, while libstdc++ doesn't (haven't checked VS).  To me this looks like a flaw in the get_time/strptime spec.  What do you think about having `parse` transform `%F` into `%Y-%m-%d`?  That should be easy enough to do (in addition to your fix).

**Comment by acolomitchi** [comment]
The fix may not be complete, I haven't checked the local_time version of parse. Also, implementation-wise, the fact that I'm break-ing a switch case prematurely doesn't taste good to me (it's like having returns in the mid of a function). Fell free to ignore the fix I proposed.

> To me this looks like a flaw in the get_time/strptime spec.

Same to me.

>  What do you think about having parse transform %F into %Y-%m-%d?

Opinion: if you decide for extra additions, you'll need to take the responsibility of documenting them. The [time_get reference](http://en.cppreference.com/w/cpp/locale/time_get/get#Parameters) doesn't specify a %F; so if you don't have "spare cycles" for the docos, one can live without %F.

Otherwise, yes, I feel is worth it. And (if I'm not grabbing your full arm when offered one finger) I'd suggest other enhancements:

- if possible, make %Z/%z to act as optionals - don't raise the error if the fmtStr specify one of them but the parsed string is missing the timezone field. Saves some CPU - most of the time, the logic will dictate one would need to test the input with both formats if one is failing (say "Try first with tz. If it fails, try without and, if it parses Ok, consider a default tz - either your local or UTC, whatever makes... mmm... business sense in the context") 

- would be nice a %something as a shortcut for "ISO8601 full datetime format" (YYYY-mm-ddTHH:MM:SS.sss{offset}?) - ummm... where did I see it? Ah, yes, PHP5 [added a `c`](http://php.net/manual/en/function.date.php)  for it.

**Comment by HowardHinnant** [comment]
Fixed with:
https://github.com/HowardHinnant/date/commit/3b3a27efe13af1f9ddb3cbecc0b562b5a84bb4d9

I stopped short of implementing the enhancements.  But that should not imply that they won't happen in the future.

**Comment by acolomitchi** [comment]
> I stopped short of implementing the enhancements.

:) :) (groan) Mate, from what I see, you haven't stopped soon enough. :) :)

You implemented the `%F` shortcut, probably by copy/paste/adjust, **before applying the fixes for #79**

As such, would I modify my test case (in the opening comment of this bug) to use "**%Q** %T" (%Q is erroneous, by everyone makes a typo now and then), I'll trip the bug again due to what is now line 4160.

Do you want me to apply the fix again or will you?

<pre>
            case 'F':
                if (command && !modified)
                {
                    // ---------- Oy! same kind of error as issue #79!
                    f.get(is, 0, is, err, &tm, b, i-1);
                    b = i+1;
                    if ((err & ios_base::failbit) == 0)
                    {
                        const CharT ymd[] = {'%', 'Y', '-', '%', 'm', '-', '%', 'd'};
                      // ---------- Oy! same kind of error as issue #79!
                       f.get(is, 0, is, err, &tm, ymd, ymd+8);
                    }
                }
                command = false;
                modified = false;
                break;
</pre>

**Comment by HowardHinnant** [comment]
`<chuckle>` no worries.  I actually ran your test code except I used `%K`. ;-)
